paderbox/transform/module_fbank.py

```diff
@@ -73,7 +88,30 @@ def ifbanks(self):
         return np.linalg.pinv(self.fbanks.T).T
 
     def __call__(self, x):
-        x = np.dot(x, self.fbanks)
+        if self.warping_fn is None:
+            x = x @ self.fbanks
+        else:
+            independent_axis = [ax if ax >= 0 else x.ndim+ax for ax in self.independent_axis]
+            assert all([ax < x.ndim-1 for ax in independent_axis])
+            size = [
+                x.shape[i] if i in independent_axis else 1
+                for i in range(x.ndim-1)
+            ]
+            fbanks = get_fbanks(
+                n_mels=self.n_mels,
+                fft_length=self.fft_length,
+                sample_rate=self.sample_rate,
+                fmin=self.fmin,
+                fmax=self.fmax,
+                warping_fn=self.warping_fn,
+                size=size,
+            ).astype(np.float32)
+            fbanks = fbanks / (fbanks.sum(axis=-1, keepdims=True) + 1e-6)
+            fbanks = fbanks.swapaxes(-2, -1)
+            if fbanks.shape[-3] == 1:
+                x = x @ fbanks.squeeze(-3)
```

Comment:
Why is a squeeze necessary? Especially on a different axis, depending on if one dimension is a singleton?

Reply:
Given an x of, e.g., shape (B,T,F) one of the following matrix multiplications may be performed
- (B,T,1,F) @ (B,T,F,N) when fbanks have shape (B,T,F,N)
- (B,T,F) @ (B,F,N) when fbanks have shape (B,1,F,N)

Reply:
Ah, now I have an idea, where the problem is. Does `np.einsum('...F,...FN->...N')` do the correct operation in all cases?

Reply:
It does the correct operations but when I timed it, it was much slower than the current approach.

I run the following timing:

```
import numpy as np

def matmul1(a, b):
    if b.shape[-3] == 1:
        a = a @ b.squeeze(-3)
    else:
        a = (a[..., None, :] @ b).squeeze(-2)
    return a

def matmul2(a, b):
    a = (a[..., None, :] @ b).squeeze(-2)

def matmul3(a, b):
    return np.einsum('...F,...FN->...N',a,b)

a = np.random.randn(32,1000,257)
b = np.random.randn(32,1,257,80)

%timeit matmul1(a,b)
%timeit matmul2(a,b)
%timeit matmul3(a,b)
```
and got

32.8 ms ± 29.5 µs per loop (mean ± std. dev. of 7 runs, 10 loops each),
93.6 ms ± 12.9 µs per loop (mean ± std. dev. of 7 runs, 10 loops each),
566 ms ± 152 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)

for matmul1, matmul2 and matmul3, respectively.

Reply:
Ah, ok, that is a big difference.
I didn't expect such a difference, I never profiled dot vs batched dot vs einsum loop.

How about a comment like the following in the source code?
```
# The following ts the same as `np.einsum('...F,...FN->...N', x, fbanks)`,
but much faster (see https://github.com/fgnt/paderbox/pull/35)
```
So no one will simplify the code in the future to the einsum code.